packages/plugin-tanstack-query/src/engine/index.ts

```diff
@@ -0,0 +1,163 @@
+ */
+export class TanStackQueryClientEngine
+	implements BaseQueryClientEngine< QueryClient, QueryClientConfig >
+{
+	private client?: QueryClient;
+	private readonly clientOptions: QueryClientConfig;
+
+	/**
+	 * Creates a new instance of TanStackQueryClientEngine.
+	 * @param { QueryClientConfig } options - Optional QueryClient instance. If not provided, a new QueryClient is created.
+	 */
+	constructor( options?: QueryClientConfig ) {
+		options = options || ( {} as QueryClientConfig );
+
+		this.clientOptions = options;
+	}
+
+	/**
+	 * Initializes a new TanStackClient instance with default options and merges them with provided options.
+	 *
+	 * @param { QueryClientConfig } options Optional TanStackClientOptions to merge with the default configuration.
+	 *
+	 * @return A new instance of TanStackClient with the merged configuration.
+	 */
+	init( options?: QueryClientConfig ): QueryClient {
+		return new QueryClient( options );
+	}
+
+	/**
+	 * Returns the TanStack QueryClient instance.
+	 * @param { QueryClientConfig } options - Generic client options (not used in this implementation).
+	 * @return The QueryClient instance.
+	 */
+	getClient( options?: QueryClientConfig ): QueryClient {
+		return this.init( options );
+	}
+
+	/**
+	 * Returns the TanStack QueryClient instance for server-side usage.
+	 *
+	 * @return The QueryClient instance.
+	 */
+	getServerClient(): QueryClient {
+		if ( ! this.client ) {
+			this.client = this.init( this.clientOptions );
+		}
+
+		return this.client;
+	}
+
+	/**
+	 * Returns the provided QueryClient instance or undefined.
+	 * This method is useful when an optional client needs to be used.
+	 * @param { QueryClient | undefined } client - A QueryClient instance or undefined.
+	 * @return The provided QueryClient instance or undefined.
+	 */
+	useClient( client?: QueryClient ): ReturnType< typeof useQueryClient > {
+		// eslint-disable-next-line react-hooks/rules-of-hooks -- This is a hook, so we need to use it in a React component.
+		return useQueryClient( client );
+	}
+
+	/**
+	 * Executes a GraphQL query using the TanStack Client instance and writes the result to the cache.
+	 * @param { Object } props An object containing:
+	 * @param { string[] } props.key - An array of strings that uniquely identifies the query in the cache.
+	 * @param { DocumentNode } props.query - A GraphQL DocumentNode or TypedDocumentNode representing the query.
+	 * @param { TQueryOptions } props.options - Optional query options compatible with TanStack's QueryOptions.
+	 * @return A promise that resolves with the query data of type TData.
+	 */
+	// @ts-expect-error TS2416: stricter generics than interface — safe in this context
+	async fetchQuery<
+		TData,
+		TQueryOptions extends FetchQueryOptions & {
+			variables?: Record< string, unknown >;
+		} = FetchQueryOptions & {
+			variables?: Record< string, unknown >;
+		},
+	>( {
+		key,
+		query,
+		options,
+	}: {
+		key: string[];
```

Comment:
I don't like that eslint/tsc didnt pick this up.

```suggestion
		name: string;
```

Reply:
Is nobody else concerned that our tooling didnt notice this? @sh4lin @ayushnirwal

Reply:
@justlevine The time when this happened TS was ignored for the fetchquery for some other reasons and because of that our tooling didn't notice this.